stackslib/src/chainstate/burn/atc.rs

```diff
@@ -0,0 +1,1403 @@
+        //
+        // (x1, y1) = (0.2, 0.75)
+        // (x2, y2) = (0.8, 0.01)
+        //
+        // to derive L, x0, and k:
+        // L = 0.8
+        // z = ln(L/y1 - 1) / ln(L/y2 - 1)
+        // x0 = (x1 - z * x2) / (1 - z)
+        // k = ln(L/y1 - 1) / (x1 - x0)
+        //
+        // The values for x0 and k were generated with the following GNU bc script:
+        // ```
+        // $ cat /tmp/variables.bc
+        // scale=32
+        // supremum=0.8   /* this is L */
+        // x1=0.2
+        // y1=0.75
+        // x2=0.8
+        // y2=0.01
+        // z=l(supremum/y1 - 1)/l(supremum/y2 -1)
+        // x0=(x1 - z * x2)/(1 - z)
+        // k=l(supremum/y1 - 1)/(x1 - x0)
+        // print "x0 = "; x0
+        // print "k = "; k
+        // ```
+        //
+        // This script evaluates to:
+        // ```
+        // $ bc -l < /tmp/variables.bc
+        // x0 = .42957690816204645842320195118064
+        // k = 11.79583008928205260028158351938437
+        // ```
+
+        let L: f64 = 0.8;
+
+        // truncated f64
+        let x0: f64 = 0.42957690816204647;
+        let k: f64 = 11.795830089282052;
+
+        // natural logarithm constant
+        let e: f64 = 2.718281828459045;
+
+        let adv = L / (1.0 + e.powf(-k * (x0 - atc)));
+        adv
+    }
+
+    #[test]
+    fn make_null_miner_lookup_table() {
+        use crate::chainstate::burn::atc::ATC_LOOKUP;
+        let mut lookup_table = Vec::with_capacity(1024);
+        for atc in 0..1024 {
+            let fatc = (atc as f64) / 1024.0;
+            let lgst_fatc = null_miner_logistic(fatc);
+            let lgst_rational = AtcRational::from_f64_unit(lgst_fatc);
+            assert_eq!(ATC_LOOKUP[atc], lgst_rational);
+            lookup_table.push(lgst_rational);
+        }
+        println!("[");
+        for lt in lookup_table.into_iter() {
+            let inner = lt.into_inner();
+            println!("   AtcRational(Uint256({:?})),", &inner.0);
+        }
+        println!("]");
+    }
+}
+
+/// Pre-calculated 1024-member lookup table for the null miner advantage function, as AtcRational
+/// fixed point integers.  The first item corresponds to the value of the function at 0.0, and the
+/// last item corresponds to the function at 1.0 - (1.0 / 1024.0).  The input to a function is the
+/// assumed total commit carryover -- the ratio between what the winning miner paid in this
+/// block-commit to the median of what they historically paid (for an epoch-defined search window
+/// size).  A value greater than 1.0 means that the miner paid all of the assumed commit
+/// carry-over, and the null miner has negligible chances of winning.  A value less than 1.0 means
+/// that the miner underpaid relative to their past performance, and the closer to 0.0 this ratio
+/// is, the more likely the null miner wins and this miner loses.
+///
+/// This table is generated with `make_null_miner_lookup_table()` above.
+pub(crate) const ATC_LOOKUP: [AtcRational; 1024] = [
```

Comment:
Can you move this definition above the test module? I think it's better to have the test module come at the end of the file.

Reply:
Sure